## 04.03 Reverse Bits

Write a program that takes a 64-bit unsigned integer and returns the 64-bit unsigned integer consisting of the bits of the input in reverse order.  For example if the input is:
1110000000000001, the output should be 
1000000000000111.

### Hint:
Use a lookup table

### Initial Remarks:

I don't really need a lookup table, but I'll try several approaches and compare them.

In [1]:
WORD_SIZE = 16
INT_SIZE = 64
WORDS_IN_INTEGER = INT_SIZE // WORD_SIZE

MASK = 0xFFFF
PRECOMPUTED=None
if not PRECOMPUTED:
    print("Creating precomputed")
    PRECOMPUTED = {i: int(bin(i)[2:].zfill(WORD_SIZE)[::-1].zfill(WORD_SIZE), 2) for i in range(2**WORD_SIZE) }
else:
    print("Precomputed exists.")
   
def solution_1(my_int):
    """Use strings"""
    bit_string = bin(my_int)[2:].zfill(64)
    rev = bit_string[::-1]
    answer = int(rev, 2)
    
    return my_int, answer, rev

def solution_2(my_int):
    """Use a hash"""
    answer, start = 0, False
    remainder = my_int
    for i in range(WORDS_IN_INTEGER):
        part = remainder & MASK
        remainder = remainder >> WORD_SIZE
        if start:
            answer = answer << WORD_SIZE
        else:
            start = True
        answer += PRECOMPUTED[part]

    bit_string = bin(my_int)[2:].zfill(INT_SIZE)
    rev = bin(answer)[2:].zfill(INT_SIZE)
    
    return my_int, answer, rev
    
def book_solution(my_int):
    """Use the book solution"""
    answer = PRECOMPUTED[my_int & MASK] << 3 * WORD_SIZE \
        | PRECOMPUTED[(my_int >> WORD_SIZE) & MASK] << 2 * WORD_SIZE \
        | PRECOMPUTED[(my_int >> WORD_SIZE * 2) & MASK] << WORD_SIZE \
        | PRECOMPUTED[(my_int >> WORD_SIZE * 3) & MASK]
    return my_int, answer, bin(answer)[2:].zfill(INT_SIZE)   

from random import randint
for solution in [solution_1, solution_2, book_solution]:
    print("\n{}".format(solution.__name__))
    print(solution(142))
    print("\n{} inverse".format(solution.__name__))
    print(solution(8142508126285856768))
    print("Time:")
    %timeit solution(randint(0, 2**INT_SIZE-1))


Creating precomputed

solution_1
(142, 8142508126285856768, '0111000100000000000000000000000000000000000000000000000000000000')

solution_1 inverse
(8142508126285856768, 142, '0000000000000000000000000000000000000000000000000000000010001110')
Time:
2.2 µs ± 27.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

solution_2
(142, 8142508126285856768, '0111000100000000000000000000000000000000000000000000000000000000')

solution_2 inverse
(8142508126285856768, 142, '0000000000000000000000000000000000000000000000000000000010001110')
Time:
3.38 µs ± 73.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

book_solution
(142, 8142508126285856768, '0111000100000000000000000000000000000000000000000000000000000000')

book_solution inverse
(8142508126285856768, 142, '0000000000000000000000000000000000000000000000000000000010001110')
Time:
2.74 µs ± 149 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


### Comparison of complexity among the three solutions

| Solution Name   | Space Complexity | Time Complexity | Execution Time in $ \mu s $  | 
| --------------- | ---------------- | --------------- | ---------------------------- |
| `solution_1`    | $ O(n) $         | $ O(n) $        | 2.20                         |
| `solution_2`    | $ O(1) $         | $ O(n) $        | 3.38                         |
| `book_solution` | $ O(1) $         | $ O(n) $        | 2.74                         |


### Design differences among the three solutions

#### Solution 1
- uses strings
- uses built-in 'bin'
- uses built-in 'int'

#### Solution 2
- using a single loop that executes four times (once for each 16-bit word)
- using a pre-computed lookup table
- using bit-wise operations on an integer instead of using strings

#### Book Solution

- very similar to Solution 2, it just unrolls the loop

### Analysis

I know that solution_1 runs faster, but I really don't understand how.  Perhaps strings and built-in functions are just really fast.

